crates/warg/tests/package.rs

```diff
@@ -1,81 +1,105 @@
-}
-
-fn execute_test(test: Test) {
-    let envelopes: Vec<Envelope<package::model::PackageRecord>> = test
-        .input
-        .into_iter()
-        .scan(None, |last, e_data| {
-            dbg!(e_data.contents.clone());
-            let key: signing::PrivateKey = e_data.key.parse().unwrap();
-            let mut record: package::model::PackageRecord = e_data.contents.try_into().unwrap();
-
-            record.prev = last.clone();
 
-            let envelope = Envelope::signed_contents(&key, record).unwrap();
+    Ok(())
+}
 
-            *last = Some(hash::HashAlgorithm::Sha256.digest(envelope.content_bytes()));
+fn validate_input(input: Vec<EnvelopeData>) -> Result<Validator> {
+    let mut validator = Validator::new();
+    for record in input.into_iter().scan(None, |last, e_data| {
+        let key: signing::PrivateKey = e_data.key.parse().unwrap();
+        let mut record: package::model::PackageRecord = e_data.contents.try_into().unwrap();
 
-            Some(envelope)
-        })
-        .collect();
+        record.prev = last.clone();
 
-    let mut validation_state = Ok(package::validate::ValidationState::Uninitialized);
+        let envelope = Envelope::signed_contents(&key, record).unwrap();
 
-    for envelope in envelopes {
-        if matches!(validation_state, Err(_)) {
-            break;
-        }
+        *last = Some(hash::HashAlgorithm::Sha256.digest(envelope.content_bytes()));
 
-        validation_state = validation_state.unwrap().validate_envelope(&envelope);
+        Some(envelope)
+    }) {
+        validator.validate(&record)?;
     }
 
-    let result = match validation_state {
-        Ok(state) => {
-            if let ValidationState::Initialized(state) = state {
-                state.entry_state.into()
-            } else {
-                panic!("Test did not initialize state. Test input must not be empty.");
-            }
-        }
-        Err(error) => PackageStateSummary::Invalid {
-            error: format!("{}", error),
-        },
+    Ok(validator)
+}
+
+fn execute_test(input_path: &Path) -> Result<()> {
+    let output_path = Path::new("./tests/package-logs/output").join(
+        input_path
+            .file_name()
+            .ok_or_else(|| anyhow!("expected a file name for test input"))?,
+    );
+    let input: Vec<EnvelopeData> =
+        serde_json::from_str(&fs::read_to_string(input_path).with_context(|| {
+            format!(
+                "failed to read input file `{path}`",
+                path = input_path.display()
+            )
+        })?)
+        .with_context(|| {
+            format!(
+                "failed to deserialize input file `{path}`",
+                path = input_path.display()
+            )
+        })?;
+
+    let output = match validate_input(input) {
+        Ok(validator) => Output::Valid(validator),
+        Err(e) => Output::Error(e.to_string()),
     };
 
-    assert_eq!(test.output, result);
-}
+    if std::env::var_os("BLESS").is_some() {
```

Comment:
I'm planning to add two more features to the test infrastructure and I'm curious how you would combine them with the changes you've made here like `BLESS`.
1. Lettings test-writers use aliases (e.g. `{alice-pub}`) in their JSON (for input and output) that are replaced with their actual values from some central file before deserialization.
2. Having the test json generate encoded protobuf that lives in the repository so that we know that a proto we encoded in the past still works instead of / in addition to knowing that the proto we encode today works.

Reply:
For #1, if it's desirable to have the aliasing syntax in the baseline, would it be possible post-process the serialized output (i.e. as a `serde_json::Value`) to do a reverse look-up of public key to alias? If that's too much, I'm fine with not using a "bless this baseline" scheme.

For #2, I would imagine a test that reads in and validates protobuf package log; we could then simply save a textual representation of the log (json-serialized protobuf envelopes?) next to it as a baseline to compare against, with perhaps the serialized contents of the validator that was used?